src/Datadog.Trace/DuckTyping/DuckType.Statics.cs

```diff
@@ -31,9 +32,51 @@ public static partial class DuckType
         private static readonly MethodInfo _methodBuilderGetToken = typeof(MethodBuilder).GetMethod("GetToken", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
 
         [DebuggerBrowsable(DebuggerBrowsableState.Never)]
-        private static ModuleBuilder _moduleBuilder = null;
+        private static readonly Dictionary<Assembly, ModuleBuilder> ActiveBuilders = new Dictionary<Assembly, ModuleBuilder>();
+
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+        private static long _assemblyCount = 0;
         [DebuggerBrowsable(DebuggerBrowsableState.Never)]
-        private static AssemblyBuilder _assemblyBuilder = null;
+        private static long _typeCount = 0;
+
+        internal static long AssemblyCount => _assemblyCount;
+
+        internal static long TypeCount => _typeCount;
+
+        /// <summary>
+        /// Gets the ModuleBuilder instance from a target type.  (.NET Framework / Non AssemblyLoadContext version)
+        /// </summary>
+        /// <param name="targetType">Target type for ducktyping</param>
+        /// <returns>ModuleBuilder instance</returns>
+        private static ModuleBuilder GetModuleBuilder(Type targetType)
+        {
+            Assembly targetAssembly = targetType.Assembly ?? typeof(DuckType).Assembly;
+
+            if (targetType.IsGenericType)
+            {
+                foreach (var type in targetType.GetGenericArguments())
+                {
+                    if (type.Assembly != targetAssembly)
+                    {
+                        var name = new AssemblyName($"DuckTypeGenericTypeAssembly.{targetType.Name}_{++_assemblyCount}");
+                        name.Version = targetAssembly.GetName().Version;
+                        var assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(name, AssemblyBuilderAccess.Run);
+                        return assemblyBuilder.DefineDynamicModule("MainModule");
```

Comment:
NIT: Could you extract this logic in a helper method, and use the same below?

Reply:
done in 1e77200